clippy_lints/src/large_stack_arrays.rs

```diff
@@ -0,0 +1,60 @@
+use rustc::hir::*;
+use rustc::lint::{LateContext, LateLintPass, LintArray, LintPass};
+use rustc::mir::interpret::{ConstValue, Scalar};
+use rustc::ty;
+use rustc::{declare_tool_lint, impl_lint_pass};
+
+use if_chain::if_chain;
+
+use crate::rustc_target::abi::LayoutOf;
+use crate::utils::span_help_and_lint;
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for local arrays that may be too large.
+    ///
+    /// **Why is this bad?** Large local arrays may cause stack overflow.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    /// ```rust,ignore
+    /// let a = [0u32; 1_000_000];
+    /// ```
+    pub LARGE_STACK_ARRAYS,
+    pedantic,
+    "allocating large arrays on stack may cause stack overflow"
+}
+
+pub struct LargeStackArrays {
+    maximum_allowed_size: u64,
+}
+
+impl LargeStackArrays {
+    #[must_use]
+    pub fn new(maximum_allowed_size: u64) -> Self {
+        Self { maximum_allowed_size }
+    }
+}
+
+impl_lint_pass!(LargeStackArrays => [LARGE_STACK_ARRAYS]);
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for LargeStackArrays {
+    fn check_expr(&mut self, cx: &LateContext<'_, '_>, expr: &Expr) {
+        if_chain! {
+            if let ExprKind::Repeat(_, _) = expr.kind;
+            if let ty::Array(element_type, cst) = cx.tables.expr_ty(expr).kind;
+            if let ConstValue::Scalar(Scalar::Raw{data: element_count, ..}) = cst.val;
```

Comment:
From the documentation of [`Scalar::Raw`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/mir/interpret/enum.Scalar.html#variant.Raw):
> The first size bytes of data are the value. Do not try to read less or more bytes than that. The remaining bytes must be 0.

To be save, this should probably be 
```suggestion
            if let ConstValue::Scalar(Scalar::Raw{data: element_count, size: 8}) = cst.val;
```